### PR TITLE
Making sure that stop is not "blocked" by the check() in HostConfig

### PR DESCRIPTION
**Pull request by jfclere**
I have seen:
04-Apr-2023 06:12:45.499 SEVERE [main] org.apache.catalina.core.ContainerBase.stopInternal A child container failed during stop
	java.util.concurrent.ExecutionException: org.apache.catalina.LifecycleException: An invalid Lifecycle transition was attempted ([before_stop]) for component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[/clusterbench]] in state [DESTROYED]
		at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
		at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
		at org.apache.catalina.core.ContainerBase.stopInternal(ContainerBase.java:992)
		at org.apache.catalina.util.LifecycleBase.stop(LifecycleBase.java:257)
		at org.apache.catalina.core.ContainerBase$StopChild.call(ContainerBase.java:1412)
		at org.apache.catalina.core.ContainerBase$StopChild.call(ContainerBase.java:1401)
		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
		at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
		at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:145)
		at org.apache.catalina.core.ContainerBase.stopInternal(ContainerBase.java:986)
		at org.apache.catalina.util.LifecycleBase.stop(LifecycleBase.java:257)
		at org.apache.catalina.core.StandardService.stopInternal(StandardService.java:497)
		at org.apache.catalina.util.LifecycleBase.stop(LifecycleBase.java:257)
		at org.apache.catalina.core.StandardServer.stopInternal(StandardServer.java:979)
		at org.apache.catalina.util.LifecycleBase.stop(LifecycleBase.java:257)
		at org.apache.catalina.startup.Catalina.stop(Catalina.java:849)
		at org.apache.catalina.startup.Catalina.start(Catalina.java:811)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:568)
		at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:345)
		at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:476)
	Caused by: org.apache.catalina.LifecycleException: An invalid Lifecycle transition was attempted ([before_stop]) for component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[/clusterbench]] in state [DESTROYED]
		at org.apache.catalina.util.LifecycleBase.invalidTransition(LifecycleBase.java:430)
		at org.apache.catalina.util.LifecycleBase.stop(LifecycleBase.java:244)
		at org.apache.catalina.core.ContainerBase$StopChild.call(ContainerBase.java:1412)
		at org.apache.catalina.core.ContainerBase$StopChild.call(ContainerBase.java:1401)
		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
		at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
		at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:145)
		at org.apache.catalina.core.ContainerBase.stopInternal(ContainerBase.java:986)
		... 20 more
04-Apr-2023 06:12:45.500 SEVERE [main] org.apache.catalina.core.ContainerBase.stopInternal A child container failed during stop
	java.util.concurrent.ExecutionException: org.apache.catalina.LifecycleException: A child container failed during stop
		at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
		at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
		at org.apache.catalina.core.ContainerBase.stopInternal(ContainerBase.java:992)
		at org.apache.catalina.util.LifecycleBase.stop(LifecycleBase.java:257)

### CODE REVIEW

**Comment by markt-asf** [comment]
This doesn't seem quite right to me. If a container has been destroyed, it is not valid to call stop(). I think we need to understand how this is happening before we start looking at potential fixes.

**Comment by jfclere** [comment]
It seems that happens when a war is removed while tomcat is stopping,

**Comment by markt-asf** [comment]
Just deleted on the file system or removed via the Manager app?

It looks like there are sorts of race possibilities between `Host` and `HostConfig`. How far do we want to go to address these?  Since `HostConfig` is a `LifecycleListener` for `Host` I think all the hooks we might need are in place.

Would it be sufficient to make `HostConfig.check()` (and `check(String)`) a NO-OP unless the associated `Host` is in state `STARTED`?

**Comment by jfclere** [comment]
Just deleted on the file system.

**Comment by jfclere** [comment]
I need to find a way to reproduce to investigate more.

**Comment by jfclere** [comment]
@markt-asf there is something weird while removing the war file and letting Tomcat undeploying the webapp, randomly Tomcat seems to reject the connection of the shutdown port:
java.net.ConnectException: Connection refused

I created 200 war files put them webapps and wait until they are started and then remove the war files and try to stop Tomcat....

**Comment by jfclere** [comment]
OK the  lot of times is because of the Thread.sleep(500); in checkResources() in HostConfig.

**Comment by jfclere** [comment]
According to a bunch of test I also have:
+++
SEVERE [main] org.apache.catalina.core.ContainerBase.removeChild Error destroying child
        org.apache.catalina.LifecycleException: An invalid Lifecycle transition was attempted ([before_destroy]) for component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[/hello-sailor]] in state [STARTED]
                at org.apache.catalina.util.LifecycleBase.invalidTransition(LifecycleBase.java:430)
                at org.apache.catalina.util.LifecycleBase.destroy(LifecycleBase.java:316)
                at org.apache.catalina.core.ContainerBase.removeChild(ContainerBase.java:826)
                at org.apache.catalina.core.ContainerBase.destroyInternal(ContainerBase.java:1033)
+++
Or 
[main] org.apache.catalina.util.LifecycleBase.destroy The destroy() method was called on component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[/hello-sailor]] after destroy() had already been called. The second call will be ignored ...

**Comment by rmaucher** [comment]
> Just deleted on the file system or removed via the Manager app?
> 
> It looks like there are sorts of race possibilities between `Host` and `HostConfig`. How far do we want to go to address these? Since `HostConfig` is a `LifecycleListener` for `Host` I think all the hooks we might need are in place.
> 
> Would it be sufficient to make `HostConfig.check()` (and `check(String)`) a NO-OP unless the associated `Host` is in state `STARTED`?

There is no periodic event unless the container is available so this is "done" already to some extent. However without a sync it is not fully enforced.
I think the PR at this point enforces it "more".

Maybe adding synchronized to ContainerBase.backgroundProcess as general "sane behavior" ? Also HostConfig.check() and HostConfig.check(String name) should sync on the host an recheck getState().isAvailable() to cover when the call comes through the manager ?

File: java/org/apache/catalina/startup/HostConfig.java [hunk 1]
```diff
@@ -154,6 +154,16 @@ public class HostConfig implements LifecycleListener {
...
+    /**
+     *  Tell that we are in check()
+     */
+    protected volatile boolean isDoingCheck = false;
+
+    /**
+     *  Tell that we are stopping (set after event beforestop).
+     */
+    protected volatile boolean isDoingStop = false;
+
```
Comment (markt-asf) [inline java/org/apache/catalina/startup/HostConfig.java:166]: Needs to correctly handle a sequence start, stop, start, stop etc.

**Comment by jfclere** [comment]
See https://github.com/apache/tomcat/commit/dce1fc833b5192ec0ccee7d4120a050cee5da8de
the exceptions are gone but now the stop() is blocked by the check(String name) but the fix is easy don't wait for removed war resources.

**Pull request closed without merging**